### DISCUSSION TITLE
Is it possible to render window/tile borders over the content?

### DISCUSSION THREAD

**Discussion by al-caveman** [qtile/qtile#3322] [Q&A]
Windows/tiles have borders.  The borders are rendered on the outside of windows.

What I want is to render borders on the inside, so borders will be hovering over window/tile contents.  Basically, I'm looking for semi-transparent 2px-wide such transparent borders hovering on the window/tile contents (not on the outside).

Why?  Because I want to have 100% pixel utilisation for my monitors.  I plan to show borders, and the status bar, only when I press the mod key.  When I release the mod key, all borders and the status bar disappear.  Having borders being rendered to hover over window contents is beneficial here as it will avoid resizing windows upon the appearance of their borders.

**Comment by m-col** [comment]
Sounds pretty cool. Unfortunately that's a non-trivial feature that I think is unlikely to make it into Qtile. However I don't think it would be impossible for a plugin to provide the feature as a keybinding without any modification of Qtile's internal code. If you wanted to try implementing it, it might be useful to look at how bspwm renders their preselection feedback, which similarly draws rectangles on a window's contents.

**Reply by al-caveman** [reply]
What makes it non-trivial?

**Reply by m-col** [reply]
Hmm, I guess that's a subjective term so probably not helpful. Maybe it wouldn't be so hard to implement but I can't say for sure. I suspect you'd need a function to enable it, and another to disable it. The 'enable' function would need to iterate through the mapped windows and use their geometry to render some rectangles (with `xcffib.xproto.RECTANGLE` and `PolyFillRectangle`) in the right places. The bit I'm not so sure about is rendering the internal borders with transparency rather than on top of the window's contents, but this is what bspwm's presel_feedback appears to do so it can be done.

**Reply by al-caveman** [reply]
Instead of iterating, can't it be always displayed, except that when mod key is released, their opacity become 100%, hence effectively not shown?

This way, the geometry of the borders will be defined only when tiles are created/resized, as usual.  The new thing will be having such borders' opacity change upon mod press/release to make them visible/invisible.

**Reply by m-col** [reply]
Yeah, probably could

**Reply by al-caveman** [reply]
Can I do this opacity trick with current Qtile? (without a custom plugin)

**Reply by m-col** [reply]
I'm afraid not. There is no code for rendering rectangles on to windows - that's what you would need to put together.

**Reply by al-caveman** [reply]
Could you please show me how are the current rectangular borders drawn in Qtile?  E.g. which part of Qtile is responsible of doing this?  How much of the drawing  of the current borders is done by Qtile, and how much is by dependencies?

**Reply by m-col** [reply]
Under X11, the X serve draws them for us. We just tell the X server the width and color that we want: https://github.com/qtile/qtile/blob/master/libqtile/backend/x11/window.py#L433-L435

Unless we want multiple borders, in which case we do need to render them a bit more manually (via the X server still though): https://github.com/qtile/qtile/blob/master/libqtile/backend/x11/window.py#L443-L461

In Wayland we render them directly using wlroots helpers: https://github.com/qtile/qtile/blob/master/libqtile/backend/wayland/output.py#L187-L203
